pkg/project/auth/locker.go

```diff
@@ -0,0 +1,55 @@
+package auth
+
+import (
+	"sync"
+)
+
+// Locker manages a pool of locks by key
+type Locker interface {
+	// GetOrCreate a lock to manage the specified key
+	GetOrCreate(key string) *sync.RWMutex
+	// Delete a lock for the given key
+	Delete(key string)
+}
+
+// locker provides
+type locker struct {
+	lock  sync.RWMutex
+	items map[string]*sync.RWMutex
+}
+
+// NewLocker creates a new locker
+func NewLocker() Locker {
+	return &locker{
+		lock:  sync.RWMutex{},
+		items: map[string]*sync.RWMutex{},
+	}
+}
+
+// GetOrCreate a lock to manage the specified key
+func (l *locker) GetOrCreate(key string) *sync.RWMutex {
+	l.lock.RLock()
+
+	keyLock := l.items[key]
+	if keyLock != nil {
+		l.lock.RUnlock()
+		return keyLock
+	}
+	l.lock.RUnlock()
+
+	// we need to create one
+	keyLock = &sync.RWMutex{}
+
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	l.items[key] = keyLock
+	return keyLock
+}
+
+// Delete the lock for the specified key
+func (l *locker) Delete(key string) {
```

Comment:
Will probably make Delete be a Release func and ensure its called after every GetOrCreate.  Will probably rename GetOrCreate to Acquire. 

Sent from my iPhone

> On Feb 11, 2015, at 8:32 AM, David Eads <notifications@github.com> wrote:
> 
> In pkg/project/auth/locker.go:
> 
> > +		return keyLock
> > +	}
> > +	l.lock.RUnlock()
> > +
> > +	// we need to create one
> > +	keyLock = &sync.RWMutex{}
> > +
> > +	l.lock.Lock()
> > +	defer l.lock.Unlock()
> > +
> > +	l.items[key] = keyLock
> > +	return keyLock
> > +}
> > +
> > +// Delete the lock for the specified key
> > +func (l *locker) Delete(key string) {
> Is there a WeakHashMap construct in Go? That makes a Locker easier to reason about since you never have to worry about deletion logic. Once the references are gone, the object can be GC'ed. Otherwise, you'll need a sort of ref counter prior to delete to protect from multiple GetOrCreates racing with a Delete on a different thread and ending up with different locks on each thread.
> 
> —
> Reply to this email directly or view it on GitHub.
>